documentation/general/dotnet-run-file.md

```diff
@@ -0,0 +1,224 @@
+# `dotnet run file.cs`
+
+This is a proposal for extending the dotnet CLI to allow running C# source files with no need for an explicit backing project.
+We call these *file-based programs* (as opposed to *project-based programs*).
+
+```ps1
+dotnet run file.cs
+```
+
+> [!NOTE]
+> This document describes the ideal final state, but the feature will be implemented in [stages](#stages).
+
+## Motivation
+
+File-based programs
+- should be a viable alternative to using PowerShell/bash scripts in .NET repos, and
+- lower the entry barrier for new customers.
+
+## Guiding principle
+
+The overarching guiding principle is that file-based programs have a simple and reliable [grow up](#grow-up) story to project-based programs.
+Previous file-based approaches like scripting are a variant of C# and as such have no simple and reliable grow up story.
+
+## Implicit project file
+
+The [guiding principle](#guiding-principle) implies that we can think of file-based programs as having an implicit project file.
+
+The implicit project file is the default project that would be created by running `dotnet new console`.
+This means that the behavior of `dotnet run file.cs` can change between SDK versions if the `dotnet new console` template changes.
+In the future we can consider supporting more SDKs like the Web SDK.
+
+## Grow up
+
+When file-based programs reach an inflection point where build customizations in a project file are needed,
+a single CLI command can be executed to generate a project file.
+In fact, this command simply materializes the [implicit project file](#implicit-project-file) on disk.
```

Comment:
What does a grown-up file look like in the case of multiple entry points? Does the generated file <Compile Remove="..."> all other entrypoints?

Reply:
> Does the generated file `<Compile Remove="...">` all other entrypoints?

No. I thought best would be to move the file to project subdirectories. I will clarify.